synapse/handlers/register.py

```diff
@@ -236,8 +245,10 @@ async def register_user(
         else:
             # autogen a sequential user ID
             fail_count = 0
-            user = None
-            while not user:
+            # If a default display name is not given, generate one.
+            generate_display_name = default_display_name is None
+            # This breaks on successful registration *or* errors after 10 failures.
+            while True:
```

Comment:
Change this to a loop which breaks simplifies the types of `user` (it is now always a `UserID`).